crates/kitsune_p2p/kitsune_p2p/src/gossip/simple_bloom.rs

```diff
@@ -0,0 +1,548 @@
+
+        // this value needs to be somewhat frequent to support send timing
+        let loop_check_interval_ms = std::cmp::max(send_interval_ms / 3, 100);
+
+        let gossip = this.clone();
+        metric_task(async move {
+            loop {
+                tokio::time::sleep(std::time::Duration::from_millis(
+                    loop_check_interval_ms as u64,
+                ))
+                .await;
+
+                if let GossipIterationResult::Close = gossip.run_one_iteration().await {
+                    tracing::warn!("gossip loop ending");
+                    break;
+                }
+            }
+
+            KitsuneResult::Ok(())
+        });
+
+        this
+    }
+
+    async fn run_one_iteration(&self) -> GossipIterationResult {
+        // # Step 1 - check state
+        //   - if closed, send GossipIterationResult::Close
+        //   - if not ready, exit early
+        let sync_and_initiate = match self.step_1_check() {
+            CheckResult::Close => return GossipIterationResult::Close,
+            CheckResult::NotReady => return GossipIterationResult::Good,
+            CheckResult::SyncAndInitiate => true,
+            CheckResult::SkipSyncAndInitiate => false,
+        };
+
+        if sync_and_initiate {
+            // # Step 2 - run a local sync, updating bloom / data_map / key_set
+            match self.step_2_local_sync().await {
+                Err(_) => return GossipIterationResult::Close,
+                Ok(false) => return GossipIterationResult::Good,
+                Ok(true) => (),
+            }
+
+            // # Step 3 - check target for initiation
+            //   - if we don't have a current initiation target, pick one
+            //   - send the initiate message
+            match self.step_3_initiate().await {
+                Err(_) => return GossipIterationResult::Close,
+                Ok(false) => return GossipIterationResult::Good,
+                Ok(true) => (),
+            }
+        }
+
+        // # Step 4 - loop on incoming/outgoing data in parallel
+        //   - if processing incoming data is slow we want to keep
+        //     sending outgoing data at appropriate times
+        //   - if we get a "finished" chunk from our initaite target,
+        //     clear the initiate target
+        //   - if we get through all incoming/outgoing, move on
+        //   - if we take > gossip_interval, move on
+        match self.step_4_com_loop().await {
+            Err(_) => return GossipIterationResult::Close,
+            Ok(false) => return GossipIterationResult::Good,
+            Ok(true) => (),
+        }
+
+        GossipIterationResult::Good
+    }
+
+    fn step_1_check(&self) -> CheckResult {
+        match self.0.share_mut(|i, _| {
+            // first, if we don't have any local agents, there's
+            // no point in doing any gossip logic
+            if i.local_agents.is_empty() {
+                return Ok(CheckResult::NotReady);
+            }
```

Comment:
Is this because we are only gossiping with local agents??

Reply:
No, this means no agents have joined the networking module yet - i.e. we have a conductor with no cells!